### DISCUSSION TITLE
Empty table being called (probably simple)

### DISCUSSION THREAD

**Discussion by MosheMendel18** [zauberzeug/nicegui#1280] [Q&A]
### Question

An empty table gets tossed to the top of the screen, but I never call `ui.table()` without putting in all the data.
Here's what the ui looks like: 
![image](https://github.com/zauberzeug/nicegui/assets/136655032/aac4eb9d-0fe0-4712-96d0-10ebcd11e841)

The title, image, button, and table are all in a card within a tab interface.
```py
with ui.tabs() as tabs:
    ui.tab('h', label='Home', icon='home')
    ui.tab('a', label='Path', icon='route')
    ui.tab('d', label='Devices', icon = 'devices')
with ui.tab_panels(tabs, value='h').classes('w-full'):
    with ui.tab_panel('h'):
        ui.label('Main Content')
        for r2 in ro_devs: #ro_devs is a list of integer indexes
            with ui.card().tight():
                with ui.row():
                    ui.image(dL[r2]['icon']).classes('w-16') 
                    with ui.button(dN[r2], on_click =lambda:ui.notify('Hi!', close_button='OK')):
                        badge = ui.badge('Analytics', color = 'red').props('floating')
            with ui.card_section():
                just_get(r2) #the table is drawn here
```
The `just_get()` function creates the table (after pulling data from the device whose index is provided). 
```py
@ui.refreshable
def just_get(index : int):
    mapkeys = [key for key in dMap[index]]
    outputs = list(map(lambda x: dMap[index][x]['code'],mapkeys))
    d = tinytuya.OutletDevice(dId[index], cid=nId[index],local_key= dLocalKeys[index], parent=gw)
    try:
        payload = d.generate_payload(tinytuya.DP_QUERY)
        d.send(payload)
        data = d.receive()
        decoded_outputs = list(zip(outputs,data['dps'].values()))
        rows = []
        card_df = pd.DataFrame({'Measure': [],' Value' :[]})
        for p in decoded_outputs:
            card_df = pd.concat([card_df,pd.DataFrame({'Measure' : [p[0]], 'value' : [p[1]]})],ignore_index=True)   
        table = ui.table(columns=[{'name': col, 'label': col, 'field': col} for col in card_df.columns],rows=card_df.to_dict('records'), row_key='Measure').classes('w-full')
    except Exception as e:
        print(e)
```
So I'm not sure how the table gets initiated without any data. I've also tried replacing the call to `just_get(r2)` with a `just_get.refresh(r2)`, but that doesn't work either. 

Any help or advice would be appreciated.

**Comment by falkoschindler** [comment]
Hi @MosheMendel18,
It's always helpful to provide a [minimum reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example). It helps us and yourself to pinpoint where the problem is coming from.
Since your code example is not executable, I can only guess: It looks like you're iterating over multiple `ro_devs`, causing multiple tables to be created. And if one `decoded_outputs` is empty, you'll see "No data available" in this table.

**Reply by MosheMendel18** [reply]
You're absolutely right about the minimum reproducible example. In this case however, it's not entirely possible, because it involves having a real gateway, real device, Tuya iot dev account with API key and client ID.As for `ro_devs`, it's an array with only one element, that element being the index of the Temperature and Humidity Sensor. If you print it, it will look like this: 
```
[1]
```

I can however post the json file creating the various arrays used here.
[devicesjson.txt](https://github.com/zauberzeug/nicegui/files/12197296/devicesjson.txt)
Here's the code I use to read that file, get the index of the gateway, and create the arrays: 
```py
#First turn the .txt file to a .json file
devicesFileObject = open(f'{os.getcwd()}\\devices.json')
devicesJsonContent = devicesFileObject.read()
dL = json.loads(devicesJsonContent)
dN =[i['name'] for i in dL]
dId = [i['id'] for i in dL]
dLocalKeys = [i['key'] for i in dL]
dMacs = [i['mac'] for i in dL]
dMap = [i['mapping'] for i in dL]
nId = [i['uuid'] for i in dL]
try:
        searchString = 'gateway'
        result = list(map(lambda s: searchString in str(s).casefold(), dN))
        gw_index = result.index(True)
except Exception as e:
        print("No Gateways found!")
gw = tinytuya.Device(
        dev_id=dId[gw_index],
        address='Auto',      # Or set to 'Auto' to auto-discover IP address
        local_key=dLocalKeys[gw_index],
        persist = True,
        version=3.4)
ro_devs = [i for i,n in enumerate(dMacs) if n !='' and i != gw_index] #if it has a mac it's real ; real outlet devices
```

One thing I do notice though is that the label is not updating (notice how the label in the picture above is `MainContent**NEW**`, even though the ui.label is `MainContent`). I have seen that this happens to NiceGUI when there is an error in code, where it holds the last working version of the code. Are there any other syntax errors? Thanks for taking the time anyways.

**Comment by rodja** [comment]
> You're absolutely right about the minimum reproducible example. In this case however, it's not entirely possible, because it involves having a real gateway, real device, Tuya...

I do not understand why you would need these for a minimal example. Could you try to stepwise remove code until the error disappears? Often when encountering such errors it's really helpful to investigate it by "destroying" the app. Some parts may not be removable but can be replaced with a simpler mock. For example instead of loading data from the network, use a simple string as input etc.
